include/filewriter/filewriter_impl.hpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * @file filewriter_impl.hpp
+ * @author Jeffin Sam
+ */
+#ifndef MLBOARD_FILE_WRITER_IMPL_HPP
+#define MLBOARD_FILE_WRITER_IMPL_HPP
+
+#include "filewriter.hpp"
+
+namespace mlboard {
+
+
+fileWriter::fileWriter(std::string logdir,
+                       int maxQueueSize,
+                        std::size_t flushsec)
+{
+  const auto p1 = std::chrono::system_clock::now();
+  std::string currentTime =
+      std::to_string(std::chrono::duration_cast<std::chrono::seconds>(
+      p1.time_since_epoch()).count());
+  this->logdir = logdir + "/events.out.tfevents." + currentTime + ".v2";
+  close_ = true;
+  this->flushsec = flushsec;
+  size_t &maxSize = this->q.MaxSize();
+  maxSize = maxQueueSize;
+  thread_ = new std::thread(&fileWriter::writeSummary, this);
+  nexttime =
+      std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  outfile.open(this->logdir, std::fstream::out |
+      std::ios::trunc | std::ios::binary);
+}
+
+void fileWriter::writeSummary()
+{
+  // this is a thread that will continously write summary one by one into file
+  while (true)
+  {
+    //  Break the loop if eveything is done
+    if (!(close_ || q.size() != 0)) break;
+    std::time_t timenow =
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    if (timenow >= nexttime)
+    {
+      // wait for queue only if queue has any element or else it will
+      // forever
+      if (q.size() == 0) continue;
```

Comment:
I think we are crosschecking it in q.pop() function right? Why are we validating once again over here? Have I missed something?

https://github.com/mlpack/mlboard/pull/1/files#diff-e5e0dbb8fc2e0677bbee3d2aefd78abcR15

Reply:
if we remove the check this code will fail

```
int main()
{
    fileWriter f1("jeffin");
    f1.close();
}
```

reason is -> the event writer thread will try to pop an element but since there are 0 element it will wait 

And the notify will only work when a element is popped in ( If element is popped in then it will notify all the thread which are waiting, but since nothing is popped in it will be waiting infinitely , and hence this check

Another way of dealing this is closing  or destroying shared queue object which was intialized  in constructor, but I thought this would be safer rather than destroying since there could be elements in it :) 

Hope I cleared it, Let me know